### PR TITLE
Adding Network Interface

### PR DESCRIPTION
**Pull request by vtnerd**
This adds a C++ virtual method interface between the Posix/Rest APIs and the routing layer. Eventually the boost::future continuation code should change for the new routing_v2, and will (likely) have an interface that can be given a type-erased callback.

There are two disabled tests with this commit:
- [MAID-657](https://maidsafe.atlassian.net/browse/MAID-657)
- [MAID-658](https://maidsafe.atlassian.net/browse/MAID-658)

I'd like to get this reviewed/tested before looking at those other issues which are little more complicated to fix.

### CODE REVIEW

File: include/maidsafe/nfs/detail/async_result.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*  Copyright 2014 MaidSafe.net limited
```
Comment (dirvine) [inline include/maidsafe/nfs/detail/async_result.h]: This should probably be 2015 now (just for future ref)

**Comment by dirvine** [comment]
Looks like very good quality again Lee, I Cannot see any issues here at all, only query is windows stability perhaps with boost::future/thread bus as Qa passed this then I am cool to merge.

**Comment by vtnerd** [comment]
Are you referring to the issues that Niall found with boost::wait_for_all and windows?

**Comment by dirvine** [comment]
Yes, just in case of creep really.

**Pull request merged**